sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -328,6 +328,21 @@ trait Nondeterministic extends Expression {
   protected def evalInternal(input: InternalRow): Any
 }
 
+/**
+ * A stateful nondeterministic expression. These expressions contain state
+ * that is not stored in the parameter list.
+ */
+trait StatefulNondeterministic extends Nondeterministic {
```

Comment:
In hive stateful and deterministic are orthogonal. If we wanna add this new trait, I think it's time to figure it out the semantic. Shall we have a new trait called `Stateful`, or add an assumption that stateful functions must be nondeterministic?

Reply:
from Hive doc
```
A stateful UDF is considered to be non-deterministic, irrespective of what deterministic() returns.
```
This is corrected.

Reply:
Maybe we can just call it `Stateful` while still extending `Nondeterministic`, and in the doc we say that stateful expressions imply it's nondeterministic.